src/frontend/src/flows/addDevice/manage/pollForTentativeDevice.ts

```diff
@@ -55,110 +67,64 @@ const pageContent = (userNumber: bigint) => {
   });
 };
 
+export const pollForTentativeDevicePage = (
+  props: Parameters<typeof pollForTentativeDeviceTemplate>[0],
+  container?: HTMLElement
+): void => {
+  const contain =
+    container ?? (document.getElementById("pageContent") as HTMLElement);
+  render(pollForTentativeDeviceTemplate(props), contain);
+};
+
 /**
  * Polls for a tentative device to be added and shows instructions on how to continue the device registration process on the new device.
  * @param userNumber anchor of the authenticated user
  * @param connection authenticated II connection
  */
 export const pollForTentativeDevice = async (
   userNumber: bigint,
-  connection: AuthenticatedConnection
-): Promise<void> => {
-  const [timestamp, tentativeDevice] = await withLoader(async () =>
-    Promise.all([
-      connection.enterDeviceRegistrationMode(),
-      getTentativeDevice(await connection.getAnchorInfo()),
-    ])
+  connection: AuthenticatedConnection,
+  endTimestamp: Timestamp
+): Promise<DeviceData | "timeout" | "canceled"> => {
+  const countdown = AsyncCountdown.fromNanos(endTimestamp);
+
+  // Show the page with the option to cancel
+  const page = new Promise<"canceled">((resolve) =>
+    pollForTentativeDevicePage({
+      userNumber,
+      cancel: () => {
+        countdown.stop();
+        resolve("canceled");
+      },
+      remaining: countdown.remainingFormattedAsync(),
+    })
   );
 
-  if (tentativeDevice) {
-    // directly show the verification screen if the tentative device already exists
-    await verifyTentativeDevice({
-      connection,
-      alias: tentativeDevice.alias,
-      endTimestamp: timestamp,
-    });
-    return;
-  } else {
-    renderPollForTentativeDevicePage(userNumber);
-    await init(userNumber, connection, timestamp);
-  }
-};
+  // Poll repeatedly
+  const polling = poll(userNumber, connection, () => countdown.hasStopped());
 
-export const renderPollForTentativeDevicePage = (userNumber: bigint): void => {
-  const container = document.getElementById("pageContent") as HTMLElement;
-  render(pageContent(userNumber), container);
+  return await Promise.race([page, polling]);
 };
 
-const poll = (
+const poll = async (
   userNumber: bigint,
   connection: AuthenticatedConnection,
   shouldStop: () => boolean
-): Promise<DeviceData | null> => {
-  return connection.getAnchorInfo().then((response) => {
+): Promise<DeviceData | "timeout"> => {
+  for (;;) {
     if (shouldStop()) {
-      return null;
+      return "timeout";
     }
-    const tentativeDevice = getTentativeDevice(response);
-    if (tentativeDevice) {
+
+    const anchorInfo = await connection.getAnchorInfo();
+    const tentativeDevice =
+      anchorInfo.device_registration[0]?.tentative_device[0];
```

Comment:
this - `anchorInfo.device_registration[0]?.tentative_device[0]` - appears twice in this PR, maybe worth a utils?

Reply:
There was a helper around this but I found it more confusing than just inlining the definition; if you feel strongly I'm happy to re-add one!